Library/Homebrew/cask/lib/hbc/installer.rb

```diff
@@ -420,10 +437,46 @@ def zap
       purge_caskroom_path
     end
 
+    def backup_path
+      return nil if @cask.staged_path.nil?
+      Pathname.new "#{@cask.staged_path}.upgrading"
+    end
+
+    def backup_metadata_path
+      return nil if @cask.metadata_versioned_path.nil?
+      Pathname.new "#{@cask.metadata_versioned_path}.upgrading"
+    end
+
+    def version_is_latest?
+      @cask.versions.include?("latest")
+    end
+
     def gain_permissions_remove(path)
       Utils.gain_permissions_remove(path, command: @command)
     end
 
+    def purge_backed_up_versioned_files
+      ohai "Purging files for version #{@cask.version} of Cask #{@cask}"
+
+      # versioned staged distribution
+      gain_permissions_remove(backup_path) if !backup_path.nil? && backup_path.exist?
+
+      # Homebrew-Cask metadata
+      if backup_metadata_path.respond_to?(:children) &&
+         backup_metadata_path.exist? &&
+         backup_metadata_path.children.each do |subdir|
+           unless PERSISTENT_METADATA_SUBDIRS.include?(subdir.basename)
+             gain_permissions_remove(subdir)
+           end
+         end
+      end
+      backup_metadata_path.rmdir_if_possible
+      @cask.metadata_master_container_path.rmdir_if_possible unless upgrade?
```

Comment:
The `unless upgrade?` here doesn't make sense, because this method is only called when upgrading.